virtwho/config.py

```diff
@@ -674,40 +337,1002 @@ def parse_dests_from_dict(dict_to_parse,
+        # Any item not in this set but in this ConfigSection is validated
+        self._unvalidated_keys = set()
+        self._invalid_keys = set()
+        # Holds all properties for this section
+        self._values = {}
+        # Validation messages from last run
+        self.validation_messages = []
+
+        self.validation_methods = OrderedDict()
+        self._destinations = {}
+        self._required_keys = set()
+        self._missing_required_keys = set()
+
+        # Add section defaults
+        for key, value in self.defaults.items():
+            self._values[key] = value
+        self._update_state()
+
+    def iteritems(self):
+        for key in self:
+            yield (key, self._values[key])
+
+    def items(self):
+        return util.DictItemsIter(self)
+
+    def __iter__(self):
+        return iter(self._values.keys())
+
+    def _update_state(self):
+        if len(self._unvalidated_keys) > 0:
+            self.state = ValidationState.NEEDS_VALIDATION
+        elif len(self._invalid_keys) > 0 or len(self._values) == 0 or len(
+                self._missing_required_keys) > 0:
+            self.state = ValidationState.INVALID
+        else:
+            self.state = ValidationState.VALID
+
+    def __delitem__(self, key):
+        if key in self:
+            if self.has_default(key):
+                return
+            del self._values[key]
+            if key in self._unvalidated_keys:
+                self._unvalidated_keys.remove(key)
+            if key in self._invalid_keys:
+                self._invalid_keys.remove(key)
+            self._update_state()
+        else:
+            raise KeyError('Unable to delete nonexistant property "%s"' % key)
+
+    def __contains__(self, key):
+        return key in self._values
+
+    def __setitem__(self, key, value):
+        if key not in self or (key in self and self[key] != value):
+            self._unvalidated_keys.add(key)
+            self._values[key] = value
+            self._update_state()
+        elif key not in self._unvalidated_keys:
+            self._unvalidated_keys.add(key)
+            self._update_state()
+
+    def __len__(self):
+        return len(self._values)
+
+    def __getitem__(self, key):
+        if key in self._values:
+            return self._values[key]
+        elif key in self.defaults:
+            return self.defaults[key]
+        else:
+            raise KeyError('{0} not in {1}'.format(key, self.name))
+
+    def _pre_validate(self):
+        """
+        Steps necessary to do before evaluation 
+        """
+        if len(self._unvalidated_keys) == 0:
+            self.validation_messages.append(('warning', 'No values provided in: %s' % self.name))
+        else:
+            for default_key in self.defaults.keys():
+                if default_key not in self._unvalidated_keys:
+                    self.validation_messages.append(
+                        (
+                            'warning',
```

Comment:
Do we need to worry about internationalization at all?

Reply:
No sir!